wgctrl-rs/src/device.rs

```diff
@@ -103,6 +108,121 @@ pub struct DeviceInfo {
     pub(crate) __cant_construct_me: (),
 }
 
+type RawInterfaceName = [c_char; libc::IFNAMSIZ];
+
+/// The name of a Wireguard interface device.
+#[derive(PartialEq, Eq, Clone, Copy)]
+pub struct InterfaceName(RawInterfaceName);
+
+impl InterfaceName {
+    /// Creates a new [InterfaceName](Self).
+    pub fn new(name: &str) -> Result<Self, InvalidInterfaceName> {
```

Comment:
Thoughts on instead implementing this as `TryFrom` instead?

```rust
impl TryFrom<&str> for InterfaceName { ... }
```

That way all these methods that currently take a `interface: &str` can instead look like:

```rust
fn do_a_thing<I: TryInto<InterfaceName>>(interface: I)
```

which I think might make it easier to convert all cases where an interface is accepted as a `&str`.

Reply:
I think that's reasonable. Ill convert it over and look at replacing more cases of `interface &str`

Reply:
@mcginty I moved it to a `TryFrom<&str>` implementation, but due to a lot of the methods returning `io::Result<()>`, you can't use it as a generic bound. Since when you perform the conversion in the function, it has no idea you're passing it the `&str` implementation so it is unable to convert `InvalidInterfaceType` into `io::Error`.

Reply:
Ah yeah you'd have to do something a bit more verbose like

```rust
pub fn get_by_name<T: TryInto<InterfaceName, Error = InvalidInterfaceName>>(
    name: T,
) -> Result<DeviceInfo, io::Error> {
    let name: InterfaceName = name.try_into()?;
    ...
}
```

Which is a bit much for every signature that takes that as an argument. I'd say let's keep it how you have it now and see if the code that uses `wgctrl-rs` becomes unwieldy before making the function signatures that messy :).